### ISSUE TITLE
Cannot list configrepo-generated Pipelines as Upstream, blocking Fetch Artifact configuration

### ISSUE THREAD

**Issue by chrisb2244** [gocd/gocd#10667]
##### Issue Type
One of
- Bug Report
- Feature enhancement

##### Summary

Fetch Artifact tasks require Pipeline, Stage, Job and Source inputs, all of which can be free strings (the dropdown completion is helpful, but not required, so config-driven pipelines can be entered here with care).
However, you cannot fetch an artifact from a pipeline that is not considered Upstream.
Attempting to add the config-driven pipeline as a Material is impossible (at least with the GUI) because although the pipeline name can be entered as a free string, "Upstream Stage" is a required value and cannot be chosen (dropdown is empty).

###### Basic environment details

* Go Version: `22.1.0`
* JAVA Version: `17.0.2`
* OS: `Windows 10`

##### Steps to Reproduce

1. Generate a pipeline that produces some artifact (text file, whatever) via configuration repository (I'm using the YAML config plugin)
2. Add a new non-config-driven pipeline
3. Add a "Fetch Artifact" task and choose GoCD as the type
4. Fill out the configuration fields to match the pipeline from step 1.
5. Click save (see image 1)
6. Go to pipeline materials, add "Another Pipeline"
7. Enter the name of pipeline from step 1.
8. Click save (see image 2).

##### Expected Results

Ideally be able to add the "Upstream Stage" as a string rather than exclusively enum/dropdown.

##### Actual Results

Attempting to add Fetch Artifact step:
![image](https://user-images.githubusercontent.com/7201433/183007491-23e8138e-989e-4b6f-8844-326643ef7361.png)

Unable to add Upstream Pipeline:
![image](https://user-images.githubusercontent.com/7201433/183007536-8afd57f2-e069-4de4-849c-9ede15a598c2.png)

##### Possible Fix

Make Upstream Stage a string?
The code using this seems to be here: https://github.com/gocd/gocd/blob/f309c478a3e7b4d5c6abfc602260b304f028ae67/config/config-api/src/main/java/com/thoughtworks/go/config/materials/dependency/DependencyMaterialConfig.java#L247

but I think the part that would need changing is probably somewhere else (GUI-specific?), I couldn't find it.

##### Any other info

This is a useful tool to enable manual testing for pipeline setup, YAML configurations, and artifact creation/consumption, but it probably isn't critical to fix if complicated, because presumably the YAML configuration can be used to generate the appropriate links between pipelines and artifact specifications in a production scenario.

**Comment by chadlwilson** [comment]
I'm not 100% sure of the reason to restrict the suggestions to not include config repository pipelines, however the UI code that defines this is at

https://github.com/gocd/gocd/blob/5c7c6c226dc2caa1fe729f5142536634912e795a/server/src/main/webapp/WEB-INF/rails/webpack/views/pages/pipelines/non_scm_material_fields.tsx#L124-L148

I guess it might need to be changed to an `<AutocompleteField>` with a custom `SuggestionsProvider` which is populated from the pipeline's stages, if it has been found in the cache.

**Comment by arvindsv** [comment]
> I'm not 100% sure of the reason to restrict the suggestions to not include config repository pipelines

I believe that is by design. The expectation is that a config repository is "optional". It can go away at some point. I mean:

1. Config repository brings in pipeline_A.
2. Config XML defines pipeline_B which uses pipeline_A as upstream
3. Server restarts and the config repository is unreachable
4. Server cannot start because the config XML is invalid, and references a non-existent pipeline

This is also mentioned as the second point in the "Limitations" section in https://github.com/gocd/gocd/issues/1133#issuecomment-109785886. Also mentioned as the first point in the "Behavior and assumptions" section in https://github.com/gocd/gocd/issues/1133#issue-77045258

I believe and am fairly certain that this isn't a GUI-only restriction. I expect that it'll fail even if the config is edited manually.

**Comment by chadlwilson** [comment]
Hmm, good point - thanks for all the context.

On the other hand, you can have two different config repositories with one referring to an upstream in the other. In the case that the upstream is removed or config repo unreachable your server will load without those pipelines until the config repos parse successfully, or the upstream is removed.

I guess the difference here is that the ones directly defined on the server become "blocking" whereas ones inside config repos are treating as "pointers"/"partials" and there is generally a non-blocking path forward to keep using the other pipelines.

**Comment by arvindsv** [comment]
That's right. The direct ones are special, because of the blocking nature.
